userspace/libsinsp/event.cpp

```diff
@@ -1402,6 +1404,75 @@ Json::Value sinsp_evt::get_param_as_json(uint32_t id, OUT const char** resolved_
 	return ret;
 }
 
+std::string sinsp_evt::get_cwd(uint32_t id, sinsp_threadinfo *tinfo)
+{
+	std::string cwd = tinfo->get_cwd();
+
+	const ppm_param_info* param_info = &m_info->params[id];
+
+	// If it's a regular FSPATH, just return the thread's CWD
+	if (param_info->type != PT_FSRELPATH)
+	{
+		ASSERT(param_info->type == PT_FSPATH);
+		return cwd;
+	}
+
+	uint64_t dirfd_id = (uint64_t)param_info->info;
+	const ppm_param_info* dir_param_info = &(m_info->params[dirfd_id]);
+
+	// Ensure the index points to an actual FD
+	if (dir_param_info->type != PT_FD)
+	{
+		ASSERT(dir_param_info->type == PT_FD);
+		return cwd;
+	}
+
+	const sinsp_evt_param* dir_param = &m_params[dirfd_id];
+	const int64_t dirfd = *(int64_t*)dir_param->m_val;
+
+	// If the FD is special value PPM_AT_FDCWD, just use CWD
+	if (dirfd == PPM_AT_FDCWD)
+	{
+		return cwd;
+	}
+
+	// If the previous param is a fd with a value other than AT_FDCWD,
+	// use such value as the current working directory
+	sinsp_fdinfo_t* dir_fdinfo = tinfo->get_fd(dirfd);
+	if (!dir_fdinfo)
+	{
+#ifndef _WIN32
+		// Sad day; we don't have the directory in the tinfo's fd cache.
+		// Must manually look it up so we can resolve filenames correctly.
+		char proc_path[PATH_MAX];
+		char dirfd_path[PATH_MAX];
+		int ret;
+		snprintf(proc_path,
+		         sizeof(proc_path),
+		         "/proc/%lld/fd/%lld",
+		         (long long)tinfo->m_pid,
+		         (long long)dirfd);
+
+		ret = readlink(proc_path, dirfd_path, sizeof(dirfd_path) - 1);
+		if (ret < 0)
+		{
+			return cwd;
+		}
+		dirfd_path[ret] = '\0';
+		std::string rel_path_base = dirfd_path;
+		sanitize_string(rel_path_base);
+		rel_path_base.append("/");
+		return rel_path_base;
+#else
+		return cwd;
+#endif
```

Comment:
@nathan-b, I know you're sick of my nitpicking at this point, but this should really live in libscap and be wrapped in both compile time checks (`#ifdef HAS_CAPTURE` instead of `#ifndef _WIN32`; while both are wrong, HAS_CAPTURE is slightly less wrong, as it won't try to access /proc e.g. on macOS) and runtime checks (`if (handle->m_mode != SCAP_MODE_LIVE) { return NULL; }` or something to that effect; we don't want to look up live process fds when reading a capture)